### PR TITLE
ARROW-9122: [C++] Properly handle sliced arrays in ascii_lower, ascii_upper kernels

### PR DESCRIPTION
**Pull request by wesm**
We must both take into account the referenced range of the data buffer and shift the value offsets if the array slice offset is nonzero.

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/scalar_string_test.cc [hunk 1]
```diff
@@ -56,12 +56,14 @@ TYPED_TEST(TestStringKernels, AsciiLength) {
                    "[3, null, 0, 1]");
 }
 
-TYPED_TEST(TestStringKernels, DISABLED_AsciiUpper) {
+TYPED_TEST(TestStringKernels, AsciiUpper) {
+  this->CheckUnary("ascii_upper", "[]", this->string_type(), "[]");
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string_test.cc:60]: Does this implicitly test sliced arrays?

Comment (wesm) [reply cpp/src/arrow/compute/kernels/scalar_string_test.cc:60]: It does now -- @bkietz added this in ARROW-971

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 2]
```diff
@@ -41,16 +41,50 @@ struct AsciiLength {
 
 using TransformFunc = std::function<void(const uint8_t*, int64_t, uint8_t*)>;
 
+template <typename T>
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:44]:
```suggestion
// Transform a buffer of offsets to one which begins with 0 and has equivalent value lengths
template <typename T>
```

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 3]
```diff
@@ -41,16 +41,50 @@ struct AsciiLength {
...
+  T* out_offsets = reinterpret_cast<T*>((*out)->mutable_data());
+  T first_offset = *input_offsets;
+  for (int64_t i = 0; i < length; ++i) {
+    *out_offsets++ = input_offsets[i] - first_offset;
+  }
+  *out_offsets = input_offsets[length] - first_offset;
+  return Status::OK();
+}
+
+template <typename Type>
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:58]:
```suggestion
// Apply `transform` to input character data- this function cannot change the length
// of any string value in the input.
template <typename Type>
```

**Comment by wesm** [comment]
+1, merge on green

**Comment by wesm** [comment]
the test failure is an s390x flake

**Pull request closed without merging**